### ISSUE TITLE
Zebra ZPL 4-inch/300dpi/Direct Thermal driver is not able to save a default media size

### ISSUE THREAD

**Issue by zdohnal** [michaelrsweet/lprint#112]
I've installed my lab label printer with lprint and tried to set the correct label size - 4x6" - however the default size always gets rewritten after I save the changes and click into printer's web page in lprint and then back to 'Media' or 'Printing Defaults'.

It looks to be connected to the 'Media' and "Printing Defaults" pages - I've check that when I create printer, /etc/lprint.conf contains the correct media, however when I go to the mentioned pages, custom size is showed - 4x4.04.

Would you mind looking into it? lprint 1.2.0, pappl 1.4.2, Fedora 39

**Comment by michaelrsweet** [comment]
The printer might be reporting a label length of 4.04 inches. Can you attach a debug log?

**Comment by michaelrsweet** [comment]
Also, most of the Zebra printers have a "label calibration" process that will measure the installed label length.

**Comment by zdohnal** [comment]
Thank you for looking into it! I will add the debug log after New year - I have this on my list...

**Comment by zdohnal** [comment]
[Here](https://github.com/michaelrsweet/lprint/files/13938865/zebra-lprint.log.txt) is the lprint log. I describe the workflow I did:

1. install printer in lprint
2. go to printing defaults (here is incorrect size) - 
![screenshot](https://github.com/michaelrsweet/lprint/assets/17822565/0d62fd1a-b4b7-47c4-a6b7-84bf95006f7b)
3. change the media size and type, save - 
![screenshot](https://github.com/michaelrsweet/lprint/assets/17822565/c6e26115-52b1-4f41-bc12-6529d9ff41b7)
4. go to printing defaults (size is correct) - 
![screenshot](https://github.com/michaelrsweet/lprint/assets/17822565/952d1bfa-4c9d-41d6-8948-9c71a5bea6fa)
5. go to the printer main page and then to printing defaults - size is again incorrect - 
![screenshot](https://github.com/michaelrsweet/lprint/assets/17822565/27eab8bc-9fc1-4f19-8730-44f59f3c255e)

**Comment by michaelrsweet** [comment]
Are you sure you have a 300dpi Zebra printer? The length (1210) would be about 6" at 203dpi.

It does look like I might need to round the label sizes but let's fix the resolution first.

**Comment by zdohnal** [comment]
ok, bad on me - I suspected 300dpi and I was not able to find any settings on the machine web ui with info relevant to me, but [user manual](https://www.zebra.com/content/dam/zebra_new_ia/en-us/solutions-verticals/product/Printers/Desktop%20Printers/G-Series%20GK%20Desktop%20Printers/GENERAL/spec-sheet/gk420-tech-spec-sheet-en-us.pdf) says it clearly - I have bad resolution.

And changing the driver did the trick.

=============================================

To sum it up for future users - consult your machine supported resolution and PDL with your machine manual.

**Comment by zdohnal** [comment]
Hmm... looks like the default size changes more and more - now 4x5.8... maybe it counts the width of label and gives the actual width based on how perferct I tear the media?

**Comment by zdohnal** [comment]
The default size is continuously changing to the smaller and smaller width - probably I have something broken with media type I use - I've tried continuous + labels, web + labels, web + continuous labels, continuous + continous labels (anything with 'mark' produced many labels - the printer was not able to see mark). Media provided to the printer is like [this one](https://www.amazon.com/4x6-Direct-Thermal-Labels-Printers/dp/B079PWB1HX).

**Comment by michaelrsweet** [comment]
I added the "gap" method to the ZPL driver; it should be able to recognize it.

Also, I updated the media size detection code to do some rounding and "snapping" before choosing the size. Please try the latest master code...

**Comment by zdohnal** [comment]
Ok, web labels did the trick for several test pages (maybe the width was broken from previous attempts), so I guess it was again problem on my side - I really have spend more time with label printers.


I'll close the issue - if I see it again, I'll reopen.
